File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int n=grid.size();
        int m=grid[0].size();
        vector<vector<int>> isVisited(n,vector<int>(m,0));
        queue<pair<int,int>> q;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==2){
                    q.push({i,j});
                }
            }
        }
        vector<vector<int>> temp={{1,0},{0,1},{-1,0},{0,-1}};
        while(q.size()){
            int size=q.size();
            for(int i=0;i<size;i++){
                int x=q.front().first;
                int y=q.front().second;
                q.pop();
                for(int j=0;j<temp.size();j++){
                    int x_=x+temp[j][0];
                    int y_=y+temp[j][1];
                    if(x_>=0 && y_>=0 && x_<n && y_<m && isVisited[x_][y_]==0 && grid[x_][y_]==1){
                        isVisited[x_][y_]=isVisited[x][y]+1;
                        q.push({x_,y_});
                    }
                }
            }
        }
        int ans=0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(isVisited[i][j]==0 && grid[i][j]==1) return -1;
                ans=max(ans,isVisited[i][j]);
            }
        }
        return ans;
    }
};

File: 1036-rotting-oranges/rotting-oranges.java
class Solution {
    class Pair{
        int i,j;
        Pair(int i,int j){
            this.i=i;
            this.j=j;
        }
    }
    public int orangesRotting(int[][] grid) {
        int m=grid.length,n=grid[0].length;
        Queue<Pair>q=new LinkedList<>();
        int vis[][]=new int[m][n];
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]==2){
                    q.add(new Pair(i,j));
                }
            }
        }
        while(q.size()>0){
            Pair p=q.remove();
            int i=p.i,j=p.j;
            if(i+1<m && vis[i+1][j]==0 && grid[i+1][j]==1){
                q.add(new Pair(i+1,j));
                vis[i+1][j]=vis[i][j]+1;
            } 
            if(j+1<n && vis[i][j+1]==0 && grid[i][j+1]==1){
                q.add(new Pair(i,j+1));
                vis[i][j+1]=vis[i][j]+1;
            } 
            if(i-1>=0 && vis[i-1][j]==0 && grid[i-1][j]==1){
                q.add(new Pair(i-1,j));
                vis[i-1][j]=vis[i][j]+1;
            } 
             if(j-1>=0 && vis[i][j-1]==0  && grid[i][j-1]==1){
                q.add(new Pair(i,j-1));
                vis[i][j-1]=vis[i][j]+1;
            }  
        }
        int max=0;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(vis[i][j]==0 && grid[i][j]==1) return -1;
                max=Math.max(max,vis[i][j]);
            }
        }
        return max;
        
        
    }
}